docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+
+### Risks and Mitigations
+
+One risk for this process is that the user experience is slightly convoluted. A user (or cloud provider)
+will need to create these fields for any machine template which will be used to scale from zero.
+This association might not seem obvious at first as the autoscaler will only see the information
+throught the MachineSets and MachineDeployments.
+
+Creating clear documentation about the flow information, and the action of the cluster autoscaler
+will be the first line of mitigating the confusion around this process. Additionally, adding an
+example in the test Docker provider will help to clarify usage for cloud providers.
+
+## Alternatives
+
+An alternative approach to reconciling the resource hints from the machine template to MachineSets
+and MachineDeployments would be for the cluster autoscaler to directly read the machine template
+through the `infrastructureRef` that exists within those objects. This would require the autoscaler
+to query for the machine templates, as well as the required permissions for those objects.
+
+Another alternative would be for cloud providers to dynamically provide this information at run time.
+This may be an approach to consider in the future, but it would still require a method to expose
+the data at the MachineSet and MachineDeployment level. In effect this might look like an
+automated method of achieving the same action as this proposal.
+
+A much larger alternative would be to create a new custom resource that would act as an autoscaling
+abstraction. This new resource would be accessed by both the cluster autoscaler and the Cluster API
+controllers, as well as potentially another operator to own its lifecycle. This approach would
+provide the cleanest separation between the components, and allow for future features in a contained
+environment. The downside is that this approach requires the most engineering and design work to
+accomplish.
+
+## Upgrade Strategy
+
+As this field is optional, it should not negatively affect upgrades. That said, care should be taken
+to ensure that this field is copied during any object upgrade as its absence will create unexpected
+behavior for end users.
+
+## Additional Details
+
+### Test Plan
+
+The cluster autoscaler tests for Cluster API integration do not currently exist outside of the downstream
+testing done by Red Hat on the OpenShift platform. There have talks over the last year to improve this
+situation, but it is slow moving currently.
+
+The end goal for testing is to contribute the scale from zero tests that currently exist for OpenShift
+to the wider Kubernetes community. This will not be possible until the testing infrastructure around
+the cluster autoscaler and Cluster API have resolved more.
+
+## Implementation History
+
+- [X] 06/10/2021: Proposed idea in an issue or [community meeting]
+- [ ] MM/DD/YYYY: Compile a Google Doc following the CAEP template (link here)
```

Comment:
You can reference the previous proposal as a preliminary work, or remove entirely this line